keras_nlp/layers/random_word_deletion.py

```diff
@@ -0,0 +1,187 @@
+    ... )
+    >>> augmenter(["I like to fly kites, do you?",
+    ...     "Can we go fly some kites later?"])
+    <tf.Tensor: shape=(2,), dtype=string, numpy=array([b'I fly kites, do you?',
+    b'Can we fly kites later?'], dtype=object)>
+
+    Augment first, then batch the dataset.
+    >>> tf.random.get_global_generator().reset_from_seed(30)
+    >>> tf.random.set_seed(30)
+    >>> inputs = ["I like to fly kites, do you?",
+    ...     "Can we go fly some kites later?"]
+    >>> augmenter = keras_nlp.layers.RandomWordDeletion(
+    ...     probability = 0.6,
+    ...     max_deletions = 3,
+    ... )
+    >>> ds = tf.data.Dataset.from_tensor_slices(inputs)
+    >>> ds = ds.map(augmenter)
+    >>> ds = ds.apply(tf.data.experimental.dense_to_ragged_batch(3))
+    >>> ds.take(1).get_single_element()
+    <tf.Tensor: shape=(2,), dtype=string, numpy=array([b'fly kites, do you?',
+    b'we go some kites'], dtype=object)>
+
+    Batch the inputs and then Augment.
+    >>> tf.random.get_global_generator().reset_from_seed(30)
+    >>> tf.random.set_seed(30)
+    >>> inputs = ["I like to fly kites, do you?",
+    ...     "Can we go fly some kites later?"]
+    >>> augmenter = keras_nlp.layers.RandomWordDeletion(
+    ...     probability = 0.6,
+    ...     max_deletions = 3,
+    ... )
+    >>> ds = tf.data.Dataset.from_tensor_slices(inputs)
+    >>> ds = ds.batch(3).map(augmenter)
+    >>> ds.take(1).get_single_element()
+    <tf.Tensor: shape=(2,), dtype=string, numpy=array([b'fly kites, do you?',
+    b'we go some kites'], dtype=object)>
+    """
+
+    def __init__(self, probability, max_deletions, **kwargs):
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.int32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_integer and dtype != tf.string:
+                raise ValueError(
+                    "Output dtype must be an integer type or a string. "
+                    f"Received: dtype={dtype}"
+                )
+
+        super().__init__(**kwargs)
+        self.probability = probability
+        self.max_deletions = max_deletions
+
+    def call(self, inputs):
+        """Augments input by randomly deleting words.
+
+        Args:
+            inputs: A tensor or nested tensor of strings to augment.
+
+        Returns:
+            A tensor or nested tensor of augmented strings.
+        """
+
+        isString = False
+        if isinstance(inputs, str):
+            inputs = [inputs]
+            isString = True
+
+        if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+            inputs = tf.convert_to_tensor(inputs)
+            inputs = tf.cast(inputs, tf.string)
+
+        def _map_fn(inputs):
+            scalar_input = inputs.shape.rank == 0
+            if scalar_input:
+                inputs = tf.expand_dims(inputs, 0)
+
+            ragged_words = tf.strings.split(inputs)
```

Comment:
I'm a little worried about how we are doing whitespace splitting here. There's two downsides with this approach.

 - It's destructive. Newlines, tabs, consecutive spaces will all get replaces with a single space.
 - It will group punctuation along with words and delete them both. So for an input "Are you OK?" It would delete "OK?" as a single token I think.

Do we want that behavior? Maybe in the context of EDA normal usage it is fine.

There might be an option to do something fancier here. Use `tf_text.regex_split` to split on whitespace and punctuation, but still keep the split characters. Use `tf.strings.regex_full_match` to exclude whitespace and punctation from deletion. Something like...

```
ragged_words = tf_text.regex_split(inputs, WHITESPACE_AND_PUNCTUATION_REGEX, WHITESPACE_AND_PUNCTUATION_REGEX)

positions_flat = tf.range(tf.size(ragged_words.flat_values))
positions = ragged_words.with_flat_values(positions_flat)
seperator_positions = tf.strings.regex_full_match(ragged_words, WHITESPACE_AND_PUNCTUATION_REGEX)
positions = tf.ragged.boolean_mask(positions_flat, separator_positions == False)
```

Reply:
Thanks @mattdangerw working on this rn, just to confirm in the last line it should be 
`tf.ragged.boolean_mask(positions, separator_positions == False)` instead right?

Reply:
@mattdangerw This method also does this weird thing where it preserves the whitespaces for the deleted texts too, as you can see in this [demo](https://colab.research.google.com/gist/aflah02/e9672690129b5193851883115a21784d/demodeletionwithregex.ipynb) the whitespaces in front of "like" and "to" are not preserved making the output worse. This does satisfy the criteria of preserving spaces albeit in a very weird way